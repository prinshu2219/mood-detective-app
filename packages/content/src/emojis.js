export const moodEmojis = {
    happy: {
        primary: '😊',
        variants: ['😀', '😁', '😄', '😃', '🙂', '😍', '🥰', '😋', '🤗', '😎'],
    },
    sad: {
        primary: '😢',
        variants: ['😭', '😔', '😞', '😥', '😪', '🥺', '😿', '😢', '😰', '😨'],
    },
    angry: {
        primary: '😠',
        variants: ['😡', '🤬', '😤', '😾', '💢', '😠', '😣', '😖', '😫', '😩'],
    },
    neutral: {
        primary: '😐',
        variants: ['😑', '😶', '🤔', '🤨', '😐', '😯', '😦', '😧', '😮', '😲'],
    },
};
export const characterEmojis = {
    luna: '🕵️‍♀️',
    tom: '😊',
    maya: '😢',
    rex: '😠',
};
export const gameEmojis = {
    star: '⭐',
    trophy: '🏆',
    medal: '🥇',
    sparkle: '✨',
    check: '✅',
    cross: '❌',
    arrow: '➡️',
    heart: '❤️',
};
export const backgroundEmojis = {
    cloud: '☁️',
    sun: '☀️',
    rainbow: '🌈',
    star: '⭐',
    sparkle: '✨',
    flower: '🌸',
    leaf: '🍃',
};
// Helper function to get random emoji from a category
export function getRandomEmoji(category) {
    const emojis = moodEmojis[category];
    const allEmojis = [emojis.primary, ...emojis.variants];
    return allEmojis[Math.floor(Math.random() * allEmojis.length)];
}
// Helper function to get emoji for mood label
export function getMoodEmoji(mood) {
    const moodKey = mood.toLowerCase();
    return moodEmojis[moodKey]?.primary || '😐';
}
