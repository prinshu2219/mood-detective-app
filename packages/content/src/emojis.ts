export const moodEmojis = {
  happy: {
    primary: '😊',
    variants: ['😀', '😁', '😄', '😃', '🙂', '😍', '🥰', '😋', '🤗', '😎'],
  },
  sad: {
    primary: '😢',
    variants: ['😭', '😔', '😞', '😥', '😪', '🥺', '😿', '😢', '😰', '😨'],
  },
  angry: {
    primary: '😠',
    variants: ['😡', '🤬', '😤', '😾', '💢', '😠', '😣', '😖', '😫', '😩'],
  },
  neutral: {
    primary: '😐',
    variants: ['😑', '😶', '🤔', '🤨', '😐', '😯', '😦', '😧', '😮', '😲'],
  },
};

export const characterEmojis = {
  luna: '🕵️‍♀️',
  tom: '😊',
  maya: '😢',
  rex: '😠',
};

export const gameEmojis = {
  star: '⭐',
  trophy: '🏆',
  medal: '🥇',
  sparkle: '✨',
  check: '✅',
  cross: '❌',
  arrow: '➡️',
  heart: '❤️',
};

export const backgroundEmojis = {
  cloud: '☁️',
  sun: '☀️',
  rainbow: '🌈',
  star: '⭐',
  sparkle: '✨',
  flower: '🌸',
  leaf: '🍃',
};

// Helper function to get random emoji from a category
export function getRandomEmoji(category: keyof typeof moodEmojis): string {
  const emojis = moodEmojis[category];
  const allEmojis = [emojis.primary, ...emojis.variants];
  return allEmojis[Math.floor(Math.random() * allEmojis.length)];
}

// Helper function to get emoji for mood label
export function getMoodEmoji(
  mood: 'HAPPY' | 'SAD' | 'ANGRY' | 'NEUTRAL'
): string {
  const moodKey = mood.toLowerCase() as keyof typeof moodEmojis;
  return moodEmojis[moodKey]?.primary || '😐';
}
