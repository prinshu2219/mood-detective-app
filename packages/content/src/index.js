import strings from './strings.en.json';
export * from './emojis';
export { strings };
// Helper function to get nested content with type safety
export function getContent(path) {
    return path.split('.').reduce((obj, key) => obj?.[key], strings);
}
// Helper function to replace placeholders in strings
export function formatString(template, replacements) {
    return template.replace(/\{\{(\w+)\}\}/g, (match, key) => {
        return replacements[key]?.toString() || match;
    });
}
