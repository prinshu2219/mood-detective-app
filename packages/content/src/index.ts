import strings from './strings.en.json';
export * from './emojis';

export { strings };
export type { Strings } from './types';

// Helper function to get nested content with type safety
export function getContent(path: string): any {
  return path.split('.').reduce((obj: any, key) => obj?.[key], strings);
}

// Helper function to replace placeholders in strings
export function formatString(
  template: string,
  replacements: Record<string, string | number>
): string {
  return template.replace(/\{\{(\w+)\}\}/g, (match, key) => {
    return replacements[key]?.toString() || match;
  });
}
