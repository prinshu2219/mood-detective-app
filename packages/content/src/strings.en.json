{
  "welcome": {
    "title": "Mood Detective",
    "subtitle": "Can a computer read feelings? Join Detective Luna to find out!",
    "startButton": "Start Adventure",
    "description": "A friendly journey for 6th graders."
  },
  "characters": {
    "luna": {
      "name": "Detective Luna",
      "role": "Your AI Guide",
      "description": "A friendly detective who loves solving mood mysteries!",
      "dialogue": [
        "Hi there! I'm Detective Luna, and I'm here to help you understand how computers can read feelings!",
        "Did you know that computers can look at words and figure out if someone is happy, sad, or angry?",
        "It's like being a detective, but instead of looking for clues, we look for mood words!",
        "Ready to become a mood detective with me?"
      ]
    },
    "tom": {
      "name": "Tom",
      "role": "The Happy Helper",
      "description": "Always cheerful and ready to share positive vibes!",
      "dialogue": [
        "Hey! I'm Tom, and I love finding happy words everywhere!",
        "Words like 'awesome', 'fantastic', and 'love' make me smile!",
        "Can you spot the happy words in sentences?"
      ]
    },
    "maya": {
      "name": "Maya",
      "role": "The Thoughtful Friend",
      "description": "Calm and wise, helps understand neutral feelings.",
      "dialogue": [
        "Hello! I'm Maya. Sometimes feelings aren't just happy or sad.",
        "Some sentences are neutral - they don't show strong emotions.",
        "It's important to recognize when someone is just sharing information."
      ]
    },
    "rex": {
      "name": "Rex",
      "role": "The Emotion Explorer",
      "description": "Helps identify sad and angry feelings with care.",
      "dialogue": [
        "Hi! I'm Rex, and I help understand all kinds of feelings.",
        "Sometimes people feel sad or angry, and that's okay!",
        "Words like 'upset', 'angry', or 'sad' help us understand these feelings."
      ]
    }
  },
  "story": {
    "intro": [
      "Welcome to Mood Detective Academy!",
      "I'm Detective Luna, and I'm here to teach you how computers understand feelings.",
      "Just like detectives look for clues, we look for special words that tell us about moods.",
      "Are you ready to become a mood detective?"
    ],
    "howItWorks": [
      "Here's how it works:",
      "Computers read each word in a sentence.",
      "They look for happy words like 'love' and 'awesome'.",
      "They also find sad words like 'sad' and 'upset'.",
      "And angry words like 'angry' and 'hate'.",
      "Then they decide what mood the sentence shows!"
    ],
    "challenge": [
      "Now it's your turn!",
      "I'll show you sentences, and you'll guess the mood.",
      "Drag each sentence to the right mood basket.",
      "Let's see how good you are at reading feelings!"
    ]
  },
  "sentences": {
    "happy": [
      "I love ice cream!",
      "This is the best day ever!",
      "I'm so excited about the party!",
      "My new puppy is awesome!",
      "I got an A on my test!",
      "The movie was fantastic!",
      "I'm really happy today!",
      "This game is so much fun!",
      "I love playing with my friends!",
      "The weather is perfect for a picnic!"
    ],
    "sad": [
      "I lost my favorite toy.",
      "My friend moved away.",
      "I'm feeling sad today.",
      "I miss my grandma.",
      "The movie made me cry.",
      "I'm upset about the rain.",
      "I feel lonely.",
      "My pet is sick.",
      "I'm disappointed about the game.",
      "I'm worried about the test."
    ],
    "angry": [
      "I'm so angry right now!",
      "That's not fair!",
      "I hate when this happens!",
      "I'm furious about this!",
      "This is really annoying!",
      "I'm mad at my brother!",
      "This makes me so angry!",
      "I can't believe this!",
      "I'm irritated by the noise.",
      "This is the worst!"
    ],
    "neutral": [
      "The sky is blue.",
      "I have a red backpack.",
      "The book is on the table.",
      "It's raining outside.",
      "I live in a house.",
      "The cat is sleeping.",
      "I like pizza.",
      "The car is parked.",
      "The tree has green leaves.",
      "I'm wearing a shirt."
    ]
  },
  "game": {
    "instructions": "Drag each sentence to the mood basket that matches how it feels!",
    "round": "Round {{number}} of {{total}}",
    "score": "Score: {{correct}}/{{total}}",
    "correct": "Great job! That's right!",
    "incorrect": "Not quite! Let's try again.",
    "nextRound": "Next Round",
    "finish": "Finish Game",
    "congratulations": "Congratulations! You're a Mood Detective!"
  },
  "mlPeek": {
    "title": "How Computers Think",
    "subtitle": "See how different methods analyze the same sentence",
    "rulesMethod": "Rule-Based Method",
    "mlMethod": "Machine Learning Method",
    "explanation": "Both methods look at the same words but think about them differently!",
    "trySentence": "Try your own sentence:"
  },
  "certificate": {
    "title": "Mood Detective Certificate",
    "subtitle": "This certificate is awarded to",
    "achievement": "for successfully completing the Mood Detective training",
    "skills": "Skills demonstrated:",
    "skill1": "• Identifying happy, sad, angry, and neutral emotions in text",
    "skill2": "• Understanding how computers analyze feelings",
    "skill3": "• Recognizing mood words and their impact",
    "date": "Date: {{date}}",
    "signature": "Detective Luna",
    "download": "Download Certificate"
  },
  "ui": {
    "loading": "Loading...",
    "error": "Oops! Something went wrong. Please try again.",
    "back": "Back",
    "next": "Next",
    "skip": "Skip",
    "continue": "Continue",
    "start": "Start",
    "finish": "Finish",
    "retry": "Try Again"
  }
}
