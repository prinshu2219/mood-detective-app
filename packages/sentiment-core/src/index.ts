export * from './lexicons.js';
export * from './scorer.js';
