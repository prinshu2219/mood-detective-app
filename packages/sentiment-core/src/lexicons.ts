export const happy = [
  'love',
  'great',
  'awesome',
  'fantastic',
  'joy',
  'fun',
  'good',
  'glad',
  'smile',
  'yay',
];
export const sad = [
  'sad',
  'bad',
  'cry',
  'lost',
  'miss',
  'upset',
  'unhappy',
  'worst',
  'angry',
  'mad',
];
export const angry = [
  'angry',
  'furious',
  'annoyed',
  'hate',
  'grr',
  'rage',
  'yell',
  'bother',
  'irritated',
];
export const negations = [
  'not',
  'never',
  'no',
  "can't",
  'dont',
  "don't",
  "isn't",
  "wasn't",
  "won't",
];
export const intensifiers = ['very', 'super', 'really', 'so', 'extremely'];
export const emojis: Record<string, number> = {
  '😀': 2,
  '🙂': 1,
  '😊': 1,
  '😁': 2,
  '😍': 3,
  '😭': -3,
  '😢': -2,
  '😠': -2,
  '😡': -3,
  '😐': 0,
};

// Helper function to get sentiment score for a word
export function getWordSentiment(word: string): number {
  const lowerWord = word.toLowerCase();

  if (happy.includes(lowerWord)) return 2;
  if (sad.includes(lowerWord)) return -2;
  if (angry.includes(lowerWord)) return -2;

  return 0;
}

// Helper function to check if a word is a negation
export function isNegation(word: string): boolean {
  return negations.includes(word.toLowerCase());
}

// Helper function to check if a word is an intensifier
export function isIntensifier(word: string): boolean {
  return intensifiers.includes(word.toLowerCase());
}
