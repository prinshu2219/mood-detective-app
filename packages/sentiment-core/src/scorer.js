import { happy, sad, angry, negations, intensifiers, emojis } from './lexicons';
// Enhanced tokenization that handles contractions and punctuation better
function tokenize(text) {
    const normalized = text.toLowerCase().trim();
    if (!normalized)
        return [];
    // Match words, contractions (including n't), and emojis
    const tokenPattern = /([a-z]+(?:'[a-z]+)?|n't|[\u{1F300}-\u{1FAFF}])/gu;
    const tokens = Array.from(normalized.matchAll(tokenPattern)).map(match => match[0]);
    return tokens.filter(token => token.length > 0);
}
// Enhanced word weight calculation with better categorization
function getWordWeight(word) {
    // Check emojis first
    if (word in emojis) {
        return { weight: emojis[word], category: 'emoji' };
    }
    // Check sentiment categories
    if (happy.includes(word)) {
        return { weight: 2, category: 'happy' };
    }
    if (sad.includes(word)) {
        return { weight: -2, category: 'sad' };
    }
    if (angry.includes(word)) {
        return { weight: -2, category: 'angry' };
    }
    return { weight: 0, category: 'neutral' };
}
// Calculate confidence based on score magnitude and highlight count
function calculateConfidence(score, highlights) {
    const absScore = Math.abs(score);
    const highlightCount = highlights.length;
    // Base confidence on score magnitude
    let confidence = Math.min(absScore / 4, 1); // Max confidence at score of 4
    // Boost confidence if we have multiple highlights
    if (highlightCount > 1) {
        confidence = Math.min(confidence + (highlightCount * 0.1), 1);
    }
    // Reduce confidence for very low scores
    if (absScore < 1) {
        confidence *= 0.5;
    }
    return Math.round(confidence * 100) / 100; // Round to 2 decimal places
}
export function analyze(sentence) {
    const raw = sentence.trim();
    if (!raw) {
        return {
            label: 'NEUTRAL',
            score: 0,
            highlights: [],
            confidence: 0
        };
    }
    const tokens = tokenize(raw);
    let score = 0;
    const highlights = [];
    let negateWindow = 0; // counts down on every subsequent token
    let boost = 1;
    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        // Handle negations
        if (negations.includes(token)) {
            negateWindow = 3; // flip next 3 tokens
            continue;
        }
        // Handle intensifiers
        if (intensifiers.includes(token)) {
            boost = 1.5;
        }
        // Get word weight and category
        const { weight, category } = getWordWeight(token);
        if (weight !== 0) {
            let finalWeight = weight;
            let reason = category;
            // Apply negation (counts regardless of sentiment)
            const isNegated = negateWindow > 0;
            if (isNegated) {
                finalWeight = -finalWeight;
            }
            // Apply intensifier multiplier if allowed
            const nextToken = tokens[i + 1];
            const allowBoost = boost > 1 && (!isNegated || (isNegated && ['this', 'that', 'it'].includes(nextToken)));
            if (allowBoost) {
                finalWeight = Math.sign(finalWeight) * Math.ceil(Math.abs(finalWeight) * boost);
            }
            // Reason: if negated, keep 'negated' even if intensified; else if boosted use 'intensified'; else category
            reason = isNegated ? 'negated' : (allowBoost ? 'intensified' : category);
            // Reset boost after use if we encountered a sentiment word
            if (boost > 1)
                boost = 1;
            highlights.push({
                token,
                weight: finalWeight,
                reason
            });
            score += finalWeight;
        }
        else {
            // No sentiment weight; reset boost unless this token is an intensifier or a negation token itself
            if (!intensifiers.includes(token) && !negations.includes(token)) {
                boost = 1;
            }
        }
        // Decrement negation window for every token after encountering a negation
        if (negateWindow > 0) {
            negateWindow--;
        }
    }
    // Determine label based on score and highlights
    let label = 'NEUTRAL';
    if (score >= 1) {
        label = 'HAPPY';
    }
    else if (score <= -1) {
        // Check for angry words specifically
        const hasAngryWords = highlights.some(h => h.weight <= -2 && angry.includes(h.token));
        label = hasAngryWords ? 'ANGRY' : 'SAD';
    }
    else if (score < -2) {
        label = 'ANGRY';
    }
    // Calculate confidence
    const confidence = calculateConfidence(score, highlights);
    return {
        label,
        score,
        highlights,
        confidence
    };
}
// Utility function to get detailed analysis
export function analyzeDetailed(sentence) {
    const result = analyze(sentence);
    const tokens = tokenize(sentence);
    const processingSteps = tokens.map(token => {
        const { weight } = getWordWeight(token);
        return {
            token,
            weight,
            negated: negations.includes(token),
            intensified: intensifiers.includes(token)
        };
    });
    return {
        ...result,
        tokens,
        processingSteps
    };
}
