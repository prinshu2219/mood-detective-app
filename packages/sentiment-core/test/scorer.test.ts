import { describe, it, expect } from 'vitest';
import { analyze, analyzeDetailed, type ScoreResult } from '../src/scorer';

describe('Sentiment Analysis Engine', () => {
  describe('Basic Sentiment Detection', () => {
    it('should detect happy sentiment', () => {
      const result = analyze('I love ice cream!');
      expect(result.label).toBe('HAPPY');
      expect(result.score).toBeGreaterThan(0);
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'love', weight: 2 })
      );
    });

    it('should detect sad sentiment', () => {
      const result = analyze('I am feeling sad today');
      expect(result.label).toBe('SAD');
      expect(result.score).toBeLessThan(0);
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'sad', weight: -2 })
      );
    });

    it('should detect angry sentiment', () => {
      const result = analyze('I am so angry right now!');
      expect(result.label).toBe('ANGRY');
      expect(result.score).toBeLessThan(0);
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'angry',
          weight: -3,
          reason: 'intensified',
        })
      );
    });

    it('should detect neutral sentiment', () => {
      const result = analyze('The sky is blue');
      expect(result.label).toBe('NEUTRAL');
      expect(result.score).toBe(0);
      expect(result.highlights).toHaveLength(0);
    });
  });

  describe('Negation Handling', () => {
    it('should handle simple negation', () => {
      const result = analyze('I do not love this');
      expect(result.label).toBe('SAD');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: -2,
          reason: 'negated',
        })
      );
    });

    it('should handle negation window of 3 tokens', () => {
      const result = analyze('I do not really love ice cream');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: -2,
          reason: 'negated',
        })
      );
    });

    it('should reset negation after 3 tokens', () => {
      const result = analyze('I do not really want to love ice cream');
      // 'love' should not be negated as it's beyond the 3-token window
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'love', weight: 2, reason: 'happy' })
      );
    });

    it('should handle multiple negations', () => {
      const result = analyze('I do not hate this, I love it');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'hate', weight: 2, reason: 'negated' })
      );
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'love', weight: 2, reason: 'happy' })
      );
    });
  });

  describe('Intensifier Handling', () => {
    it('should handle intensifiers', () => {
      const result = analyze('I really love this');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: 3,
          reason: 'intensified',
        })
      );
    });

    it('should handle very strong intensifiers', () => {
      const result = analyze('I super love this');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: 3,
          reason: 'intensified',
        })
      );
    });

    it('should reset intensifier after use', () => {
      const result = analyze('I really love this and hate that');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: 3,
          reason: 'intensified',
        })
      );
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'hate', weight: -2, reason: 'angry' })
      );
    });
  });

  describe('Emoji Handling', () => {
    it('should detect happy emojis', () => {
      const result = analyze('I am happy ðŸ˜Š');
      expect(result.label).toBe('HAPPY');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'ðŸ˜Š', weight: 1, reason: 'emoji' })
      );
    });

    it('should detect sad emojis', () => {
      const result = analyze('I am sad ðŸ˜¢');
      expect(result.label).toBe('SAD');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'ðŸ˜¢', weight: -2, reason: 'emoji' })
      );
    });

    it('should detect angry emojis', () => {
      const result = analyze('I am angry ðŸ˜ ');
      expect(result.label).toBe('ANGRY');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'ðŸ˜ ', weight: -2, reason: 'emoji' })
      );
    });
  });

  describe('Complex Combinations', () => {
    it('should handle negation + intensifier', () => {
      const result = analyze('I do not really love this');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: -3,
          reason: 'negated',
        })
      );
    });

    it('should handle multiple sentiment words', () => {
      const result = analyze('I love this but hate that');
      expect(result.score).toBe(0); // +2 - 2 = 0
      expect(result.highlights).toHaveLength(2);
    });

    it('should handle complex sentence with mixed emotions', () => {
      const result = analyze(
        'I really love ice cream but I am sad about the weather'
      );
      expect(result.label).toBe('HAPPY'); // +3 - 2 = +1
      expect(result.highlights).toHaveLength(2);
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty string', () => {
      const result = analyze('');
      expect(result.label).toBe('NEUTRAL');
      expect(result.score).toBe(0);
      expect(result.confidence).toBe(0);
    });

    it('should handle whitespace only', () => {
      const result = analyze('   ');
      expect(result.label).toBe('NEUTRAL');
      expect(result.score).toBe(0);
      expect(result.confidence).toBe(0);
    });

    it('should handle punctuation', () => {
      const result = analyze('I love this!!!');
      expect(result.label).toBe('HAPPY');
      expect(result.highlights).toContainEqual(
        expect.objectContaining({ token: 'love', weight: 2 })
      );
    });

    it('should handle contractions', () => {
      const result = analyze("I don't love this");
      expect(result.highlights).toContainEqual(
        expect.objectContaining({
          token: 'love',
          weight: -2,
          reason: 'negated',
        })
      );
    });
  });

  describe('Confidence Calculation', () => {
    it('should calculate high confidence for strong sentiment', () => {
      const result = analyze('I really love this fantastic thing');
      expect(result.confidence).toBeGreaterThan(0.7);
    });

    it('should calculate low confidence for weak sentiment', () => {
      const result = analyze('I like this');
      expect(result.confidence).toBeLessThan(0.5);
    });

    it('should calculate medium confidence for moderate sentiment', () => {
      const result = analyze('I love this');
      expect(result.confidence).toBeGreaterThan(0.3);
      expect(result.confidence).toBeLessThan(0.7);
    });
  });

  describe('Detailed Analysis', () => {
    it('should provide detailed analysis', () => {
      const result = analyzeDetailed('I do not really love this');

      expect(result.tokens).toContain('i');
      expect(result.tokens).toContain('do');
      expect(result.tokens).toContain('not');
      expect(result.tokens).toContain('really');
      expect(result.tokens).toContain('love');
      expect(result.tokens).toContain('this');

      expect(result.processingSteps).toHaveLength(6);
      expect(result.processingSteps).toContainEqual(
        expect.objectContaining({
          token: 'not',
          negated: true,
          intensified: false,
        })
      );
      expect(result.processingSteps).toContainEqual(
        expect.objectContaining({
          token: 'really',
          negated: false,
          intensified: true,
        })
      );
    });
  });

  describe('Snapshot Tests for Top Examples', () => {
    it('should match snapshot for happy examples', () => {
      const examples = [
        'I love ice cream!',
        'This is the best day ever!',
        'I am so excited about the party!',
        'My new puppy is awesome!',
        'I got an A on my test!',
      ];

      const results = examples.map(example => ({
        sentence: example,
        result: analyze(example),
      }));

      expect(results).toMatchSnapshot();
    });

    it('should match snapshot for sad examples', () => {
      const examples = [
        'I lost my favorite toy.',
        'My friend moved away.',
        'I am feeling sad today.',
        'I miss my grandma.',
        'The movie made me cry.',
      ];

      const results = examples.map(example => ({
        sentence: example,
        result: analyze(example),
      }));

      expect(results).toMatchSnapshot();
    });

    it('should match snapshot for angry examples', () => {
      const examples = [
        'I am so angry right now!',
        'That is not fair!',
        'I hate when this happens!',
        'I am furious about this!',
        'This is really annoying!',
      ];

      const results = examples.map(example => ({
        sentence: example,
        result: analyze(example),
      }));

      expect(results).toMatchSnapshot();
    });

    it('should match snapshot for neutral examples', () => {
      const examples = [
        'The sky is blue.',
        'I have a red backpack.',
        'The book is on the table.',
        'It is raining outside.',
        'I live in a house.',
      ];

      const results = examples.map(example => ({
        sentence: example,
        result: analyze(example),
      }));

      expect(results).toMatchSnapshot();
    });
  });
});
