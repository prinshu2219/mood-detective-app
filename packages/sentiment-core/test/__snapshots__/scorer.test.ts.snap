// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sentiment Analysis Engine > Snapshot Tests for Top Examples > should match snapshot for angry examples 1`] = `
[
  {
    "result": {
      "confidence": 0.75,
      "highlights": [
        {
          "reason": "intensified",
          "token": "angry",
          "weight": -3,
        },
      ],
      "label": "ANGRY",
      "score": -3,
    },
    "sentence": "I am so angry right now!",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "That is not fair!",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "angry",
          "token": "hate",
          "weight": -2,
        },
      ],
      "label": "ANGRY",
      "score": -2,
    },
    "sentence": "I hate when this happens!",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "angry",
          "token": "furious",
          "weight": -2,
        },
      ],
      "label": "ANGRY",
      "score": -2,
    },
    "sentence": "I am furious about this!",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "This is really annoying!",
  },
]
`;

exports[`Sentiment Analysis Engine > Snapshot Tests for Top Examples > should match snapshot for happy examples 1`] = `
[
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "happy",
          "token": "love",
          "weight": 2,
        },
      ],
      "label": "HAPPY",
      "score": 2,
    },
    "sentence": "I love ice cream!",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "This is the best day ever!",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "I am so excited about the party!",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "happy",
          "token": "awesome",
          "weight": 2,
        },
      ],
      "label": "HAPPY",
      "score": 2,
    },
    "sentence": "My new puppy is awesome!",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "I got an A on my test!",
  },
]
`;

exports[`Sentiment Analysis Engine > Snapshot Tests for Top Examples > should match snapshot for neutral examples 1`] = `
[
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "The sky is blue.",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "I have a red backpack.",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "The book is on the table.",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "It is raining outside.",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "I live in a house.",
  },
]
`;

exports[`Sentiment Analysis Engine > Snapshot Tests for Top Examples > should match snapshot for sad examples 1`] = `
[
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "sad",
          "token": "lost",
          "weight": -2,
        },
      ],
      "label": "SAD",
      "score": -2,
    },
    "sentence": "I lost my favorite toy.",
  },
  {
    "result": {
      "confidence": 0,
      "highlights": [],
      "label": "NEUTRAL",
      "score": 0,
    },
    "sentence": "My friend moved away.",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "sad",
          "token": "sad",
          "weight": -2,
        },
      ],
      "label": "SAD",
      "score": -2,
    },
    "sentence": "I am feeling sad today.",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "sad",
          "token": "miss",
          "weight": -2,
        },
      ],
      "label": "SAD",
      "score": -2,
    },
    "sentence": "I miss my grandma.",
  },
  {
    "result": {
      "confidence": 0.5,
      "highlights": [
        {
          "reason": "sad",
          "token": "cry",
          "weight": -2,
        },
      ],
      "label": "SAD",
      "score": -2,
    },
    "sentence": "The movie made me cry.",
  },
]
`;
